libs/wal_decoder/src/decoder.rs

```diff
@@ -100,7 +100,7 @@ impl MetadataRecord {
                 Self::decode_xlog_record(&mut buf, decoded, next_record_lsn)?
             }
             pg_constants::RM_LOGICALMSG_ID => {
-                Self::decode_logical_message_record(&mut buf, decoded)?
+                Self::decode_logical_message_record(&mut buf, shard, decoded)?
```

Comment:
In #9895, I added shard filtering below this block, instead of propagating it into the decode methods. This was mostly motivated by the fact that multiple decode methods may return the same metadata record type which needs common shard filtering logic, but it might be worthwhile to decouple them regardless.

Consider moving the shard filtering for logical messages below as well, for consistency. Alternatively, let's collapse the `heapam` and `neonmgr` logic and move shard filtering into those decode methods (I can do that as a followup).